src/app/util/binding-table.cpp

```diff
@@ -42,30 +42,200 @@ CHIP_ERROR BindingTable::Add(const EmberBindingTableEntry & entry)
         return CHIP_ERROR_NO_MEMORY;
     }
     mBindingTable[newIndex] = entry;
-    if (mTail == EMBER_BINDING_TABLE_SIZE)
+    CHIP_ERROR error        = SaveEntryToStorage(newIndex, kNextNullIndex);
+    if (error == CHIP_NO_ERROR)
     {
-        mTail = newIndex;
-        mHead = newIndex;
+        if (mTail == kNextNullIndex)
+        {
+            error = SaveListInfo(newIndex);
+        }
+        else
+        {
+            error = SaveEntryToStorage(mTail, newIndex);
+        }
+        if (error != CHIP_NO_ERROR)
+        {
+            mStorage->SyncDeleteKeyValue(mKeyAllocator.BindingTableEntry(newIndex));
+        }
+    }
+    if (error != CHIP_NO_ERROR)
+    {
+        // Roll back
+        mBindingTable[newIndex].type = EMBER_UNUSED_BINDING;
+        return error;
     }
     else
     {
-        mNextIndex[mTail] = newIndex;
-        mTail             = newIndex;
+        if (mTail == kNextNullIndex)
+        {
+            mTail = newIndex;
+            mHead = newIndex;
+        }
+        else
+        {
+            mNextIndex[mTail]    = newIndex;
+            mNextIndex[newIndex] = kNextNullIndex;
+            mTail                = newIndex;
+        }
+
+        mSize++;
+        return CHIP_NO_ERROR;
     }
-    mSize++;
-    return CHIP_NO_ERROR;
 }
 
 const EmberBindingTableEntry & BindingTable::GetAt(uint8_t index)
 {
     return mBindingTable[index];
 }
 
-BindingTable::Iterator BindingTable::RemoveAt(Iterator iter)
+CHIP_ERROR BindingTable::SaveEntryToStorage(uint8_t index, uint8_t nextIndex)
+{
+    EmberBindingTableEntry & entry    = mBindingTable[index];
+    uint8_t buffer[kEntryStorageSize] = { 0 };
+    chip::TLV::TLVWriter writer;
```

Comment:
This is all inside `namespace chip`.  Please drop the unnecessary `chip::` prefixes from all the new code.